apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,872 @@
+        require(_canCancel(action), ERROR_CANNOT_CANCEL_ACTION);
+
+        address submitter = action.submitter;
+        require(msg.sender == submitter, ERROR_SENDER_NOT_ALLOWED);
+
+        if (action.state == ActionState.Scheduled) {
+            _unlockBalance(submitter, setting.collateralAmount);
+        }
+        action.state = ActionState.Cancelled;
+        emit ActionCancelled(_actionId);
+    }
+
+    function challengeAction(uint256 _actionId, uint256 _settlementOffer, bytes _context) external authP(CHALLENGE_ROLE, arr(_actionId)) {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        require(_canChallenge(action, setting), ERROR_CANNOT_CHALLENGE_ACTION);
+        require(setting.collateralAmount >= _settlementOffer, ERROR_INVALID_SETTLEMENT_OFFER);
+
+        action.state = ActionState.Challenged;
+        _challengeBalance(action.submitter, setting.collateralAmount);
+        _createChallenge(action, msg.sender, _settlementOffer, _context, setting);
+        emit ActionChallenged(_actionId);
+    }
+
+    function settle(uint256 _actionId) external {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        Challenge storage challenge = action.challenge;
+        address submitter = action.submitter;
+        address challenger = challenge.challenger;
+
+        if (msg.sender == submitter) {
+            require(_canSettle(action, setting), ERROR_CANNOT_SETTLE_ACTION);
+        } else {
+            require(_canClaimSettlement(action, setting), ERROR_CANNOT_SETTLE_ACTION);
+        }
+
+        uint256 settlementOffer = challenge.settlementOffer;
+        uint256 collateralAmount = setting.collateralAmount;
+
+        // The settlement offer was already checked to be up-to the collateral amount
+        // However, we cap it to collateral amount to double check
+        uint256 unchallengedAmount = settlementOffer >= collateralAmount ? collateralAmount : (collateralAmount - settlementOffer);
+        uint256 slashedAmount = collateralAmount - unchallengedAmount;
+
+        challenge.state = ChallengeState.Settled;
+        _unchallengeBalance(submitter, unchallengedAmount);
+        _slashBalance(submitter, challenger, slashedAmount);
+        _returnArbitratorFees(challenge);
+        emit ActionSettled(_actionId);
+    }
+
+    function disputeChallenge(uint256 _actionId) external {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        require(_canDispute(action, setting), ERROR_CANNOT_DISPUTE_ACTION);
+        require(msg.sender == action.submitter, ERROR_SENDER_NOT_ALLOWED);
+
+        Challenge storage challenge = action.challenge;
+        uint256 disputeId = _createDispute(action, setting);
+        challenge.state = ChallengeState.Disputed;
+        challenge.disputeId = disputeId;
+        disputes[disputeId].actionId = _actionId;
+        emit ActionDisputed(_actionId);
```

Comment:
Let's log the `disputeId` and `arbitrator` here as well to allow having a link to the dispute just by looking at logs